packages/type-safe-api/src/construct/integrations/s3.ts

```diff
@@ -0,0 +1,89 @@
+/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0 */
+import { IRole, Role, ServicePrincipal } from "aws-cdk-lib/aws-iam";
+import { IBucket } from "aws-cdk-lib/aws-s3";
+import {
+  ApiGatewayIntegration,
+  Integration,
+  IntegrationRenderProps,
+} from "./integration";
+import { generateCorsResponseParameters } from "../prepare-spec-event-handler/prepare-spec";
+import { bucketInvocationUri } from "../spec/utils";
+
+export interface S3IntegrationProps {
+  /**
+   * The S3 bucket to be invoked on integration
+   */
+  readonly bucket: IBucket;
+
+  /**
+   * The IAM role to be used to grant permissions to the API Gateway to invoke the S3 bucket
+   * @default - a new role will be created
+   */
+  readonly role?: IRole;
+
+  /**
+   * The path override for the integration
+   * @default - integration path is used
+   * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/integrating-api-with-aws-services-s3.html
+   */
+  readonly pathOverride?: string;
+}
+
+/**
+ * An S3 integration
+ */
+export class S3Integration extends Integration {
+  private readonly bucket: IBucket;
+  private readonly role: IRole;
+  private readonly pathOverride?: string;
+
+  constructor(props: S3IntegrationProps) {
+    super();
+
+    this.bucket = props.bucket;
+
+    if (props.role) {
+      this.role = props.role;
+    } else {
+      this.role = new Role(this.bucket, "Role", {
+        assumedBy: new ServicePrincipal("apigateway.amazonaws.com"),
+      });
+    }
+
+    this.bucket.grantReadWrite(this.role, this.pathOverride ?? "*");
```

Comment:
@cogwirrel  I tried something but i'm not really familiar with custom grants and it's a bit more complicated because the render method need the role too because we need to specify the roleArn in credentials. So I'll push something but please note it's a work in progress and I'll need some feedback about it

Reply:
No worries! :) I think we can have the constructor store/create the role but then grant the permissions in the integration's `grant` method so we have the final `path` and `method` if they're not overridden by the user :)

I was thinking we could do something like this:

Constructor:

```ts
this.role = props.role ?? new Role(this.bucket, "Role", {
  assumedBy: new ServicePrincipal("apigateway.amazonaws.com"),
});
this.method = props.method;
// ...etc
```

Then override the integration `grant` method to do the granting, eg:

```ts
public grant(props: IntegrationGrantProps) {
  const path = this.path ?? props.path;
  const method = this.method ?? props.method;

  // Replace path parameters with * to grant access to arbitrary values for path parameters
  const permissionPath = path?.replace( /{[^\}]*\}/g, "*") ?? "*";

  // Grant read access for GET/HEAD/OPTIONS/TRACE, otherwise read/write
  if (["get", "head", "options", "trace"].includes(method.toLowerCase())) {
    this.bucket.grantRead(this.role, permissionPath);
  } else {
    this.bucket.grantReadWrite(this.role, permissionPath);
  }
}
```

That way we can fall back to using the `path` and `method` defined in the API for the permissions, rather than providing `*` if the user doesn't override the path :)

Reply:
Did you see my last commit with my proposal?

Reply:
I think I scoped it too much down 😂

Reply:
I just did! I like the idea of specifying the actions based on whether it's a read/write/delete :)

Haha I think you scoped down it pretty well! 😄 Just important to swap the path parameters with `*` so IAM allows access for arbitrary values there :) eg. `/pet/{petId}/details.json` becomes `/pet/*/details.json` which means at runtime someone could hit say a `/pet/scoobydoo` API and retrieve the file at `/pet/scoobydoo/details.json` :)